flink/src/main/java/org/apache/iceberg/flink/IcebergStreamWriter.java

```diff
@@ -0,0 +1,226 @@
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergStreamWriter.class);
+
+  private final String tablePath;
+  private final PartitionSpec spec;
+  private final LocationProvider locations;
+  private final Map<String, String> properties;
+  private final FileIO io;
+  private final EncryptionManager encryptionManager;
+  private final Schema schema;
+
+  private transient TaskWriter<Row> writer;
+  private transient int subTaskId;
+  private transient int attemptId;
+
+  private IcebergStreamWriter(String tablePath, PartitionSpec spec, LocationProvider locations,
+                              Map<String, String> properties, FileIO io, EncryptionManager encryptionManager,
+                              Schema schema) {
+    this.tablePath = tablePath;
+    this.spec = spec;
+    this.locations = locations;
+    this.properties = properties;
+    this.io = io;
+    this.encryptionManager = encryptionManager;
+    this.schema = schema;
+  }
+
+  @Override
+  public void open() {
+    this.subTaskId = getRuntimeContext().getIndexOfThisSubtask();
+    this.attemptId = getRuntimeContext().getAttemptNumber();
+
+    // Initialize the task writer.
+    FileFormat fileFormat = getFileFormat();
+    long targetFileSize = getTargetFileSizeBytes();
+    FileAppenderFactory<Row> appenderFactory = new FlinkFileAppenderFactory(schema, properties);
+    OutputFileFactory outputFileFactory = new OutputFileFactory(spec, fileFormat, locations, io,
+        encryptionManager, subTaskId, attemptId);
+    this.writer = TaskWriterFactory.createTaskWriter(schema, spec, appenderFactory, outputFileFactory,
+        targetFileSize, fileFormat);
+  }
+
+  @Override
+  public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+    LOG.info("Iceberg writer({}) begin preparing for checkpoint {}", toString(), checkpointId);
+    // close all open files and emit files to downstream committer operator
+    writer.close();
+
+    writer.pollCompleteFiles().forEach(this::emit);
+    LOG.info("Iceberg writer({}) completed preparing for checkpoint {}", toString(), checkpointId);
+  }
+
+  @Override
+  public void processElement(StreamRecord<Row> element) throws Exception {
+    Row value = element.getValue();
+    writer.append(value);
+
+    // Emit the data file entries to downstream committer operator if there exist any complete files.
+    writer.pollCompleteFiles().forEach(this::emit);
+  }
+
+  @Override
+  public void dispose() throws Exception {
+    super.dispose();
+    if (writer != null) {
+      writer.close();
+      writer = null;
+    }
+  }
+
+  @Override
+  public void endInput() throws IOException {
+    // For bounded stream, it may don't enable the checkpoint mechanism so we'd better to emit the remaining
+    // data files to downstream before closing the writer so that we won't miss any of them.
+    writer.close();
+    for (DataFile dataFile : writer.pollCompleteFiles()) {
+      emit(dataFile);
+    }
```

Comment:
Is it allowed to call `pollCompleteFiles()` after `close()`. IIUC, the writer is still needed when `emit`ting the remaining datafiles as well as when calling `pollCompleteFiles()`. But possibly the contract around `close` is different than I expect it to be.

Additionally, the comment here is somewhat confusing. Based on the discussion above, and based on the [task lifecycle of Flink operators outlined here](https://ci.apache.org/projects/flink/flink-docs-stable/internals/task_lifecycle.html#operator-lifecycle-in-a-nutshell), it seems like it might be more appropriate to say something along the lines of "Once the `close` method of this `IcebergStreamWriter` is invoked, we'll no longer be able to emit any remaining data files downstream. To get around this, we implement the `BoundedOneInput` interface in order to finish processing and emit any remaining data before graceful shutdown".

My third and final concern is that if we're closing the `TaskWriter` here during a graceful shutdown, if we only poll for complete files, what will happen to any remaining data that's being buffered in files that are not marked as complete? IIUC, the `TaskWriter` does not keep its opened / in-process files in flink's state such that it can be recovered after restore or replayed from the last checkpoint after a fatal. If I have that correct, it seems to me that we would want to poll for simply any remaining open files and emit them, otherwise we risk data loss. Can somebody help me better understand how we avoid losing data from in-process but incomplete files during a graceful shutdown (say a user takes a savepoint in order to restart their application and deploy a new version of their code)?

Reply:
> But possibly the contract around close is different than I expect it to be.

@kbendick  It's better to rename the `close` to `flush` here, because the current writer.close is just closing all the opening data files and the writer could still accept record and write to the new opened files.  In the task writer abstraction pull request,  we've discussed [this](https://github.com/apache/iceberg/pull/1213#discussion_r459194667) and changed to the way similar with spark.  Say  once we need to get the whole completed data file list, we will `close` the writer. If there's new record to write, then  we will initialize a new TaskWriter to write it.

> what will happen to any remaining data that's being buffered in files that are not marked as complete?

Once the `writer#close` is invoked, then all the opening file handlers will flush their data to file system and close them,  finally the completed `DataFile` (s) will be emitted to downstream operator [IcebergFileCommitter](https://github.com/apache/iceberg/pull/1185/files#diff-a8f2a02024d184e3b677cdafc678a6a3R99). When a checkpoint come,  then the committer will flush its DataFile states to flink StateBackend.  So actually, we could recover the completed files from it.